projects/centre_console/src/drive_fsm.c

```diff
@@ -0,0 +1,294 @@
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_PARKING, state_set_ebrake);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_REVERSE, state_set_motorcontroller_output);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_DRIVE, state_set_motorcontroller_output);
+}
+
+FSM_STATE_TRANSITION(state_drive) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_PARKING, state_set_relays);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_REVERSE, state_switch_direction);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_NEUTRAL, state_set_relays);
+}
+
+FSM_STATE_TRANSITION(state_parking) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_REVERSE, state_set_ebrake);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_NEUTRAL, state_set_ebrake);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_DRIVE, state_set_ebrake);
+}
+
+FSM_STATE_TRANSITION(state_reverse) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_PARKING, state_set_relays);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_DRIVE, state_switch_direction);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_NEUTRAL, state_set_relays);
+}
+
+FSM_STATE_TRANSITION(state_switch_direction) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_REVERSE, state_reverse);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_DRIVE, state_drive);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+FSM_STATE_TRANSITION(state_fault) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_OPENED_DESTINATION_NEUTRAL_PARKING,
+                     state_neutral);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+static bool prv_destination_guard(const Fsm *fsm, const Event *e, void *context) {
+  DriveFsmStorage *storage = (DriveFsmStorage *)context;
+  return e->data == storage->destination;
+}
+
+FSM_STATE_TRANSITION(state_set_motorcontroller_output) {
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_DRIVE,
+                             prv_destination_guard, state_set_relays);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_REVERSE,
+                             prv_destination_guard, state_set_relays);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_NEUTRAL_PARKING,
+                             prv_destination_guard, state_neutral);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+FSM_STATE_TRANSITION(state_set_relays) {
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_CLOSED_DESTINATION_DRIVE,
+                             prv_destination_guard, state_drive);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_CLOSED_DESTINATION_REVERSE,
+                             prv_destination_guard, state_reverse);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_OPENED_DESTINATION_NEUTRAL_PARKING,
+                             prv_destination_guard, state_set_motorcontroller_output);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
```

Comment:
MCI set output then relays? I thought we had it the other way around. I remember you also had "state_switch_direction" as being the same as set_mci_output which would make sense if the order was reversed ....

Reply:
the way state transitions work is that they look up the `destination` property in the storage to detect where they need to go next. If we go from drive to set mci output, it'll then try to set the relay states, which is unnecessary because they're already closed. BTW we're changing this up a bit since we don't have motorcontroller relays anymore... they're controlled via precharge, so drive_fsm is gonna change